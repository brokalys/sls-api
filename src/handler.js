import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
