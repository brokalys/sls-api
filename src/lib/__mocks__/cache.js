export default {
  run: jest.fn((key, params, callback) => callback(params)),
  get: jest.fn(),
  set: jest.fn(),
};
