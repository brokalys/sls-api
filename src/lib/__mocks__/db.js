export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
