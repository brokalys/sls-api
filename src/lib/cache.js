import crypto from 'crypto';

import mysql from './db';

function hash(key) {
  return crypto.createHash('md5').update(JSON.stringify(key)).digest('hex');
}

export default {
  async run(key, params, callback) {
    const cached = await this.get(key, params);

    if (cached) {
      return cached;
    }

    const newData = await callback(params);
    await this.set(key, params, newData);

    return newData;
  },

  async get(key, params) {
    const response = await mysql.query({
      sql: `SELECT value FROM ${process.env.DB_CACHE_DATABASE}.cache WHERE \`key\` = ? and \`params_hash\` = ?`,
      values: [key, hash(params)],
    });

    if (response && response.length) {
      return JSON.parse(response[0].value);
    }
  },

  set(key, params, value) {
    return mysql.query({
      sql: `INSERT INTO ${process.env.DB_CACHE_DATABASE}.cache SET ?`,
      values: {
        key,
        params: JSON.stringify(params),
        params_hash: hash(params),
        value: JSON.stringify(value),
      },
    });
  },
};
