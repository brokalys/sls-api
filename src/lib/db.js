import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
