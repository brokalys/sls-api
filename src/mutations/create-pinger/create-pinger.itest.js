import { createTestClient } from 'apollo-server-testing';

import { server } from 'handler';
import cache from 'lib/cache';
import db from 'lib/db';

jest.mock('lib/cache');
jest.mock('lib/db');

describe('Mutation: createPinger', () => {
  let mutate;

  beforeEach(() => {
    cache.get.mockReset();

    const utils = createTestClient(server);
    mutate = utils.mutate;
  });

  describe('creates a pinger', () => {
    test('with region as a polygon', async () => {
      const response = await mutate({
        mutation: `
          mutation {
            createPinger(
              email: "demo@email.com"
              category: APARTMENT
              type: SELL
              price_min: 10000
              price_max: 100000
              region: "56.96715 24.09457, 56.97923 24.14125, 56.9825 24.17984, 56.95892 24.17559, 56.94571 24.14812, 56.93767 24.13181"
            )
          }
        `,
      });

      expect(response).toMatchSnapshot();
    });
  });

  test('fails creating with invalid polygon', async () => {
    const response = await mutate({
      mutation: `
        mutation {
          createPinger(
            email: "demo@email.com"
            category: APARTMENT
            type: SELL
            price_min: 10000
            price_max: 100000
            region: "very wrong, 56.95892 24.17559, 56.94571 24.14812, 56.93767 24.13181"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('fails creating a pinger if 5 already exist', async () => {
    db.query.mockImplementation(() => [
      { email: 'demo@email.com' },
      { email: 'demo@email.com' },
      { email: 'demo@email.com' },
      { email: 'demo@email.com' },
      { email: 'demo@email.com' },
    ]);

    const response = await mutate({
      mutation: `
        mutation {
          createPinger(
            email: "demo@email.com"
            category: APARTMENT
            type: SELL
            price_min: 10000
            price_max: 100000
            region: "56.992294 24.136619, 56.976394 23.995790, 56.924904 24.005336, 56.889288 24.108467, 56.932211 24.291935, 56.996502 24.245176"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('fails creating a pinger if input validation fails', async () => {
    const response = await mutate({
      mutation: `
        mutation {
          createPinger(
            email: "demo@email.com"
            category: APARTMENT
            type: SELL
            price_min: 10000
            price_max: 100000
            region: "56.992294 24.136619, 56.976394 23.995790, 56.924904 24.005336, 56.889288 24.108467, 56.932211 24.291935, 56.996502 24.245176"
            rooms_min: 10
            rooms_max: 5
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });
});
