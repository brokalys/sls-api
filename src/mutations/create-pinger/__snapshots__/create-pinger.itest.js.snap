// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation: createPinger creates a pinger with frequency set to custom \`weekly\` 1`] = `
Object {
  "data": Object {
    "createPinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createPinger creates a pinger with price_type set to custom \`sqm\` 1`] = `
Object {
  "data": Object {
    "createPinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createPinger creates a pinger with region as a polygon 1`] = `
Object {
  "data": Object {
    "createPinger": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createPinger fails creating a pinger if 5 already exist 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "BAD_USER_INPUT",
        "maxPingers": 5,
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Max amount of 5 PINGERS per email exceeded.",
      "path": Array [
        "createPinger",
      ],
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createPinger fails creating a pinger if input validation fails 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "BAD_USER_INPUT",
        "details": Array [
          Object {
            "context": Object {
              "key": "rooms_max",
              "label": "rooms_max",
              "limit": Object {
                "adjust": null,
                "ancestor": 1,
                "depth": 1,
                "display": "ref:rooms_min",
                "in": false,
                "iterables": null,
                "key": "rooms_min",
                "map": null,
                "path": Array [
                  "rooms_min",
                ],
                "root": "rooms_min",
                "separator": ".",
                "type": "value",
              },
              "value": 5,
            },
            "message": "\\"rooms_max\\" must be greater than or equal to ref:rooms_min",
            "path": Array [
              "rooms_max",
            ],
            "type": "number.min",
          },
        ],
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Input validation failed",
      "path": Array [
        "createPinger",
      ],
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutation: createPinger fails creating with invalid polygon 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "BAD_USER_INPUT",
        "details": Array [
          Object {
            "context": Object {
              "key": "region",
              "label": "region",
              "value": "very wrong, 56.95892 24.17559, 56.94571 24.14812, 56.93767 24.13181",
            },
            "message": "\\"region\\" needs to be a valid polygon",
            "path": Array [
              "region",
            ],
            "type": "string.polygon",
          },
        ],
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Input validation failed",
      "path": Array [
        "createPinger",
      ],
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
